src/ImageSharp.Web/Middleware/ImageSharpMiddleware.cs

```diff
@@ -127,81 +135,85 @@ public class ImageSharpMiddleware
                 return;
             }
 
-            // Get the correct service for the request.
-            IImageResolver resolver = this.resolvers.FirstOrDefault(r => r.Match(context));
-
-            if (resolver == null || !await resolver.IsValidRequestAsync(context, this.logger))
-            {
-                // Nothing to do. call the next delegate/middleware in the pipeline
-                await this.next(context);
-                return;
-            }
-
             // Create a cache key based on all the components of the requested url
             string uri = $"{context.Request.Host.ToString().ToLowerInvariant()}/{context.Request.PathBase.ToString().ToLowerInvariant()}/{context.Request.Path}{QueryString.Create(commands)}";
-
             string key = CacheHash.Create(uri, this.options.Configuration);
 
-            CachedInfo info = await this.cache.IsExpiredAsync(key, DateTime.UtcNow.AddDays(-this.options.MaxCacheDays));
-
-            var imageContext = new ImageContext(context, this.options);
-
-            if (!info.Expired)
+            // Prevent identical requests from running at the same time
+            // This reduces the overheads of unnecessary processing plus avoids file locks
+            using (await this.asyncKeyLock.LockAsync(key))
             {
-                // Image is a cached image. Return the correct response now.
-                await this.SendResponse(imageContext, key, info.LastModifiedUtc, null, (int)info.Length);
-                return;
-            }
+                // Get the correct service for the request.
+                IImageResolver resolver = this.resolvers.FirstOrDefault(r => r.Match(context));
 
-            // Not cached? Let's get it from the image resolver.
-            byte[] inBuffer = null;
-            byte[] outBuffer = null;
-            MemoryStream outStream = null;
-            try
-            {
-                inBuffer = await resolver.ResolveImageAsync(context, this.logger);
-                if (inBuffer == null || inBuffer.Length == 0)
+                if (resolver == null || !await resolver.IsValidRequestAsync(context, this.logger))
                 {
-                    // Log the error but let the pipeline handle the 404
-                    this.logger.LogImageResolveFailed(imageContext.GetDisplayUrl());
+                    // Nothing to do. call the next delegate/middleware in the pipeline
                     await this.next(context);
```

Comment:
not sure we should be calling `this.next()` inside the lock... I'm pretty sure that will cause all requests to become single threaded... we want to move the `await this.next(context);` to outside the lock with a flag to determine weather to call it instead.